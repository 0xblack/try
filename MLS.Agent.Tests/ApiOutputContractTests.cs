using System.Threading.Tasks;
using Assent;
using Recipes;
using WorkspaceServer.Models;
using WorkspaceServer.Models.Execution;
using WorkspaceServer.Tests;
using Xunit;
using Xunit.Abstractions;

namespace MLS.Agent.Tests
{
    public class ApiOutputContractTests : ApiViaHttpTestsBase
    {
        private readonly Configuration configuration;

        public ApiOutputContractTests(ITestOutputHelper output) : base(output)
        {
            configuration = new Configuration()
                .UsingExtension("json");

#if !DEBUG
            configuration = configuration.SetInteractive(false);
#endif
        }

        [Fact]
        public async Task The_Run_contract_for_compiling_code_has_not_been_broken()
        {
            var viewport = ViewportCode();

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode(),
                        viewport
                    }),
                activeBufferId: viewport.Id);

            var response = await CallRun(requestJson.ToJson());

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        [Fact]
        public async Task The_Run_contract_for_noncompiling_code_has_not_been_broken()
        {
            var viewport = ViewportCode("doesn't compile");

            var request = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode(),
                        viewport
                    }),
                activeBufferId: viewport.Id);

            var requestBody = request.ToJson();

            var response = await CallRun(requestBody);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        [Fact]
        public async Task The_Completions_contract_has_not_been_broken()
        {
            var viewport = ViewportCode("Console.Ou$$");

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode(),
                        viewport
                    }),
                activeBufferId: viewport.Id).ToJson();

            var response = await CallCompletion(requestJson);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        [Fact]
        public async Task The_signature_help_contract_has_not_been_broken()
        {
            var viewport = ViewportCode("Console.Write($$");

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode(),
                        viewport
                    }),
                activeBufferId: viewport.Id).ToJson();

            var response = await CallSignatureHelp(requestJson);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        [Fact]
        public async Task The_instrumentation_contract_has_not_been_broken()
        {

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode("int a = 1; int b = 2; a = 3; b = a;")
                    },
                    includeInstrumentation: true)
            ).ToJson();

            var response = await CallRun(requestJson);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        [Fact]
        public async Task The_run_contract_with_no_instrumentation_has_not_been_broken()
        {

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode("int a = 1; int b = 2; a = 3; b = a;")
                    },
                    includeInstrumentation: false)
            ).ToJson();

            var response = await CallRun(requestJson);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        private static Workspace.Buffer EntrypointCode(string mainContent = @"Console.WriteLine(Sample.Method());$$")
        {
            var input = $@"
using System;
using System.Linq;

namespace Example
{{
    public class Program
    {{
        public static void Main()
        {{
            {mainContent}
        }}       
    }}
}}".EnforceLF();
                
            MarkupTestFile.GetPosition(input, out string output, out var position);

            return new Workspace.Buffer(
                "EntrypointCode.cs",
                output,
                position ?? 0);
        }

        private static Workspace.Buffer ViewportCode(string methodContent = @"return ""Hello world!"";$$ ")
        {
            var input = $@"
using System.Collections.Generic;
using System;

namespace Example
{{
    public static class Sample
    {{
        public static object Method()
        {{
#region viewport
            {methodContent}
#endregion
        }}
    }}
}}".EnforceLF();

            MarkupTestFile.GetPosition(input, out string output, out var position);

            return new Workspace.Buffer(
                "ViewportCode.cs",
                output,
                position ?? 0);
        }
    }
}