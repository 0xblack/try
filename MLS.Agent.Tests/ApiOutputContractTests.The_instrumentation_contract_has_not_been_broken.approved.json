{
  "succeeded": true,
  "output": [],
  "exception": null,
  "diagnostics": [
    {
      "start": 1,
      "end": 14,
      "message": "EntrypointCode.cs(2,1): hidden CS8019: Unnecessary using directive.",
      "severity": 0,
      "id": "CS8019"
    },
    {
      "start": 15,
      "end": 33,
      "message": "EntrypointCode.cs(3,1): hidden CS8019: Unnecessary using directive.",
      "severity": 0,
      "id": "CS8019"
    },
    {
      "start": 1638,
      "end": 1647,
      "message": "EntrypointCode.cs(11,1509): warning CS0472: The result of the expression is always 'true' since a value of type 'int' is never equal to 'null' of type 'int?'",
      "severity": 2,
      "id": "CS0472"
    },
    {
      "start": 2174,
      "end": 2183,
      "message": "EntrypointCode.cs(11,2045): warning CS0472: The result of the expression is always 'true' since a value of type 'int' is never equal to 'null' of type 'int?'",
      "severity": 2,
      "id": "CS0472"
    },
    {
      "start": 2341,
      "end": 2350,
      "message": "EntrypointCode.cs(11,2212): warning CS0472: The result of the expression is always 'true' since a value of type 'int' is never equal to 'null' of type 'int?'",
      "severity": 2,
      "id": "CS0472"
    },
    {
      "start": 2873,
      "end": 2882,
      "message": "EntrypointCode.cs(11,2744): warning CS0472: The result of the expression is always 'true' since a value of type 'int' is never equal to 'null' of type 'int?'",
      "severity": 2,
      "id": "CS0472"
    },
    {
      "start": 3040,
      "end": 3049,
      "message": "EntrypointCode.cs(11,2911): warning CS0472: The result of the expression is always 'true' since a value of type 'int' is never equal to 'null' of type 'int?'",
      "severity": 2,
      "id": "CS0472"
    }
  ],
  "instrumentation": [
    {
      "filePosition": {
        "line": 10,
        "character": 12,
        "file": "EntrypointCode.cs"
      },
      "stackTrace": "    at Example.Program.Main()\n ",
      "locals": [],
      "parameters": [],
      "fields": [],
      "output": {
        "start": 0,
        "end": 0
      }
    },
    {
      "filePosition": {
        "line": 10,
        "character": 23,
        "file": "EntrypointCode.cs"
      },
      "stackTrace": "    at Example.Program.Main()\n ",
      "locals": [
        {
          "name": "a",
          "value": "1",
          "declaredAt": {
            "start": 146,
            "end": 147
          }
        }
      ],
      "parameters": [],
      "fields": [],
      "output": {
        "start": 0,
        "end": 0
      }
    },
    {
      "filePosition": {
        "line": 10,
        "character": 34,
        "file": "EntrypointCode.cs"
      },
      "stackTrace": "    at Example.Program.Main()\n ",
      "locals": [
        {
          "name": "b",
          "value": "2",
          "declaredAt": {
            "start": 157,
            "end": 158
          }
        },
        {
          "name": "a",
          "value": "1",
          "declaredAt": {
            "start": 146,
            "end": 147
          }
        }
      ],
      "parameters": [],
      "fields": [],
      "output": {
        "start": 0,
        "end": 0
      }
    },
    {
      "filePosition": {
        "line": 10,
        "character": 41,
        "file": "EntrypointCode.cs"
      },
      "stackTrace": "    at Example.Program.Main()\n ",
      "locals": [
        {
          "name": "b",
          "value": "2",
          "declaredAt": {
            "start": 157,
            "end": 158
          }
        },
        {
          "name": "a",
          "value": "3",
          "declaredAt": {
            "start": 146,
            "end": 147
          }
        }
      ],
      "parameters": [],
      "fields": [],
      "output": {
        "start": 0,
        "end": 0
      }
    }
  ],
  "variableLocations": [
    {
      "name": "b",
      "locations": [
        {
          "startLine": 10,
          "startColumn": 41,
          "endLine": 10,
          "endColumn": 42
        },
        {
          "startLine": 10,
          "startColumn": 27,
          "endLine": 10,
          "endColumn": 28
        }
      ],
      "declaredAt": {
        "start": 157,
        "end": 158
      }
    },
    {
      "name": "a",
      "locations": [
        {
          "startLine": 10,
          "startColumn": 34,
          "endLine": 10,
          "endColumn": 35
        },
        {
          "startLine": 10,
          "startColumn": 45,
          "endLine": 10,
          "endColumn": 46
        },
        {
          "startLine": 10,
          "startColumn": 16,
          "endLine": 10,
          "endColumn": 17
        }
      ],
      "declaredAt": {
        "start": 146,
        "end": 147
      }
    }
  ]
}